### DISCUSSION TITLE
Error PInvokeStackImbalance in runtime

### DISCUSSION THREAD

**Discussion by mknezevic** [ikvmnet/ikvm#415] [Q&A] [answered]
Hi,

I have a .NET project in which uses a dll created from a jar file using IkvmReference. The project builds successfully and the corresponding dll is created. However, during runtime I'm getting the error:

`Managed Debugging Assistant 'PInvokeStackImbalance' 
  Message=Managed Debugging Assistant 'PInvokeStackImbalance' : 'A call to PInvoke function 'IKVM.Runtime!IKVM.Runtime.JNI.JNIVM+LibJvm+Set_JNI_GetDefaultJavaVMInitArgsDelegate::Invoke' has unbalanced the stack. This is likely because the managed PInvoke signature does not match the unmanaged target signature. Check that the calling convention and parameters of the PInvoke signature match the target unmanaged signature.'`

Any thoughts?

I use:
Visual Studio 2022
.NET Framework 4.8.1
IKVM 8.6.2

The jar that I use is grobid-core version 0.7.1 https://github.com/kermitt2/grobid/tree/0.7.1

**Comment by wasabii** [comment]
Can you try downgrading a few versions? I hit this recently as well, and haven't had a chance to do much about it.

**Reply by mknezevic** [reply]
Sure! I will try and report the progress.

**Reply by AaronRobinsonMSFT** [reply, marked as answer]
My guess is this is an issue only for x86 due to incorrect calling convention.

https://github.com/ikvmnet/ikvm/blob/3024f42f1cb5d195790974c52f453fde5d867c67/src/IKVM.Runtime/JNI/JNIVM.cs#L49

If so, the solution is to adorn the delegate with [`UnmanagedFunctionPointerAttribute`](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.unmanagedfunctionpointerattribute?view=net-7.0&redirectedfrom=MSDN).

```csharp
[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
delegate void Set_JNI_GetDefaultJavaVMInitArgsDelegate(JNI_GetDefaultJavaVMInitArgsDelegate func);
```

**Reply by wasabii** [reply]
Yeah. That makes sense. I don't know why it wasn't caught during tests though. Initializing JNI is pretty well covered. That's why I was thinking it had to be a little bit deeper at least.

Unless you mean x86, as in not x64.

**Reply by wasabii** [reply]
This whole section is sort of a place holder for now, too, btw. Needed a way to do calls from JNI code into IKVM, in a few cases where OpenJDK C libraries would do so but not have a JNIEnv on hand.

End game is to have libjvm use hosting APIs to get back into .NET, thus making it possible for people to embed IKVM as they would embed a real JDK.

**Reply by wasabii** [reply]
Aaactually. I need to think through this more. I think JNI exports on Win32 are supposed to be stdcall. I didn't add JNICALL in jvm.cpp. Maybe that's the issue.

**Reply by wasabii** [reply]
So I think the Set_ methods (since I'm making them just be JNI-like methods) should be JNIEXPORT/JNICALL, so stdcall on Windows. However, the delegates to the callbacks should probably be cdecl, since they're not actually JNI-like methods, but just function pointers to implement the functionality behind the JNI methods.

**Reply by mknezevic** [reply]
Yes, the problem is related to x86 and IKVM 8.6.2. As downgrade was suggested, I tried with IKVM 8.6.0 I got different error:
`'System.TypeInitializationException' in IKVM.Runtime.dll`
With IKVM 8.5.2 no exception is thrown. I still have problems making the grobid component to work properly, but I believe this is related solely to grobid on Windows and not to IKVM.
